crates/holochain/src/core/sys_validate.rs

```diff
@@ -366,70 +366,58 @@ pub fn check_update_reference(
 }
 
 /// Validate a chain of actions with an optional starting point.
-pub fn validate_chain<'iter>(
-    mut actions: impl Iterator<Item = &'iter ActionHashed>,
-    persisted_chain_head: &Option<(ActionHash, u32)>,
+pub fn validate_chain<'iter, A: 'iter + ChainItem>(
+    mut actions: impl Iterator<Item = &'iter A>,
+    persisted_chain_head: &Option<(A::Hash, u32)>,
 ) -> SysValidationResult<()> {
     // Check the chain starts in a valid way.
     let mut last_item = match actions.next() {
-        Some(ActionHashed {
-            hash,
-            content: action,
-        }) => {
+        Some(item) => {
             match persisted_chain_head {
                 Some((prev_hash, prev_seq)) => {
-                    check_prev_action_chain(prev_hash, *prev_seq, action)
+                    check_prev_action_chain(prev_hash, *prev_seq, item)
                         .map_err(ValidationOutcome::from)?;
                 }
                 None => {
                     // If there's no persisted chain head, then the first action
-                    // must be a DNA.
-                    if !matches!(action, Action::Dna(_)) {
+                    // must have no parent.
+                    if item.prev_hash().is_some() {
```

Comment:
Feels like there's a loss of information here. Before this enforced that the head was always a DNA but now it doesn't enforce that?

Reply:
@freesig yes there is a loss of information, but that is actually a benefit. It lets us write functions that don't care about anything more than they need to care about.

I see this as a positive change:

- this rewrites it totally in terms of the chain structure and without any knowledge of the details of the different types of Actions
- we know that the DNA is the only action without a previous hash, so referring to "an action without a previous hash" is equivalent to saying "a DNA action". (Just like saying "a Holochain core dev who lives in Oregon" is equivalent to saying "@maackle", since I'm the only one.)
- if for some reason we were to add another Action without a previous hash, it would only make sense that that Action would also be a valid start to the chain, because why else would it not have a previous action?

If we were to define a new action without a previous hash, then it should also be a valid way to start the chain -- because why else would we create such an action? There's nowhere for that action to go other than the start of the chain. If we create that hypothetical new action, then this function will now work as is without modification. If we define it in terms of DNA hash like it was, then we would have to add checks for that other type here too, or else that new type would be unusable.

This is a very minor point but I'm belaboring it because I see this as a benefit of this approach. By using a trait, I was forced to rewrite this function in a way that's more in line with the logic we actually want to express, highlighting the fact that the thing that's special about the DNA action is precisely that it has no previous hash, not the fact that it's called "DNA". This trait helps keep functions on point.

Reply:
@maackle In this case and as the source chain is such a central aspect of Holochain, I would like to see another test that makes sure that only a DNA has no previous hash. Because otherwise we are losing this safety net.

Reply:
I see your point. I guess the issue is that the "test" is actually at the type level where all actions except DNA must have an a prev hash. I wonder if there is a way to make this more explicit. Instead of returning an option there is a method on the trait like `is_valid_genesis_action(&self) -> bool` and then it's very obvious if you are declaring this can be a genesis entry rather then just omitting the prev hash?

Reply:
I think the simplest thing to do to maintain the same overall level of information as in `develop` is to just add another check elsewhere that the first action is a `Dna`. That check could perhaps use a function like `is_valid_genesis_action` to give a more explicit name to the case where the Option is None. I'll do that.

Reply:
@freesig @jost-s see the latest commit which hopefully addresses this https://github.com/holochain/holochain/pull/1501/commits/9f4a9039c5b8b7d35063f64e78c0704a7a6dee44

Reply:
that looks solid to me 👍